ktor-server/ktor-server-core/jvm/src/io/ktor/application/newapi/KtorPlugin.kt

```diff
@@ -0,0 +1,536 @@
+ *
+ * @param SubjectT is a type of subject for the current feature that is being processed. See [subject] for more information.
+ *
+ * */
+public inline class Execution<SubjectT : Any>(private val context: PipelineContext<SubjectT, ApplicationCall>) {
+    /**
+     * Continues execution of the HTTP pipeline starting after the current plugin.
+     * Current plugin's code stops being executed from the moment of calling this function.
+     * */
+    public suspend fun proceed(): SubjectT = context.proceed()
+
+    /**
+     * Continues execution of the HTTP pipeline starting after the current plugin.
+     * At the same time, the [subject] of the execution will be changed.
+     * Current plugin's code stops being executed from the moment of calling this function.
+     * */
+    public suspend fun proceedWith(subectT: SubjectT): SubjectT = context.proceedWith(subectT)
+
+    /**
+     * Finishes execution of the current plugin as well as the current pipeline (see [Pipeline] for more information).
+     * */
+    public fun finish(): Unit = context.finish()
+
+    /**
+     * A subject of the current execution. A subject can be any object of some allowed type that is currently being
+     * processed.
+     *
+     * - For the stage of processing a call, there is no [subject] (i.e. [SubjectT] is [Unit])
+     * - For the stage of sending a response, the [subject] is an object that is currently being sent via HTTP protocol. ([SubjectT] can be [Any] type)
+     * - For the stage of receiving a request data, the [subject] is of type [ApplicationReceiveRequest]. Please, read [ApplicationReceiveRequest] javadoc for more information.
+     * */
+    public val subject: SubjectT get() = context.subject
+
+    /**
+     * A call that is being processed by the plugin in the current execution.
+     * */
+    public val call: ApplicationCall get() = context.call
+
+    // Useful methods
+
+    /**
+     * See [ApplicationEnvironment] for more details.
+     * */
+    public val environment: ApplicationEnvironment get() = context.application.environment
+
+    /**
+     * Configuration of the curreny application.
+     * */
+    public val configuration: ApplicationConfig get() = environment.config
+
+    /**
+     * Sets a shutdown hook. This method is useful for closing resources allocated by the feature.
+     * */
+    public fun onShutdown(block: suspend () -> Unit) {
+        GlobalScope.launch(context.coroutineContext) {
+            block()
+        }
+    }
+}
+
+/**
+ * Port of the current application. Same as in config.
+ * */
+public val ApplicationConfig.port: Int get() = propertyOrNull("ktor.deployment.port")?.getString()?.toInt() ?: 8080
+
+/**
+ * Host of the current application. Same as in config.
+ * */
+public val ApplicationConfig.host: String get() = propertyOrNull("ktor.deployment.host")?.getString() ?: "0.0.0.0"
+
+/**
+ * [Execution] for the stage of processing a call.
+ * */
+public typealias CallExecution = Execution<Unit>
```

Comment:
Same question here: could you provide an example of using these types in public API?